candle-pyo3/src/lib.rs

```diff
@@ -436,6 +439,126 @@ impl PyTensor {
+                        shape[i] as isize + index
+                    } else {
+                        index
+                    };
+
+                    if actual_index < 0 || actual_index >= shape[i] as isize {
+                        return Err(PyTypeError::new_err(format!(
+                            "index out of range for dimension at index {i} with indexer '{value}'",
+                            i = i,
+                            value = index
+                        )));
+                    }
+
+                    let indexer = Range {
+                        start: actual_index as usize,
+                        end: (actual_index as usize) + 1,
+                    };
+                    ranges.push(indexer);
+                    // We need to sqeeze all dimensions that are indexed directly
+                    to_squeeze.push(i);
+                } else {
+                    return Err(PyTypeError::new_err("unsupported index"));
+                }
+            }
+
+            // we need to transform the ranges into a tuple of indexers and apply them
+            let mut sliced_tensor = match ranges.as_slice() {
+                [a] => self.0.i((a.to_owned(),)).map_err(wrap_err)?,
+                [a, b] => self.0.i((a.to_owned(), b.to_owned())).map_err(wrap_err)?,
+                [a, b, c] => self
+                    .0
+                    .i((a.to_owned(), b.to_owned(), c.to_owned()))
+                    .map_err(wrap_err)?,
+                [a, b, c, d] => self
+                    .0
+                    .i((a.to_owned(), b.to_owned(), c.to_owned(), d.to_owned()))
+                    .map_err(wrap_err)?,
+                [a, b, c, d, e] => self
+                    .0
+                    .i((
+                        a.to_owned(),
+                        b.to_owned(),
+                        c.to_owned(),
+                        d.to_owned(),
+                        e.to_owned(),
+                    ))
+                    .map_err(wrap_err)?,
+                [a, b, c, d, e, f] => self
+                    .0
+                    .i((
+                        a.to_owned(),
+                        b.to_owned(),
+                        c.to_owned(),
+                        d.to_owned(),
+                        e.to_owned(),
+                        f.to_owned(),
+                    ))
+                    .map_err(wrap_err)?,
+                [a, b, c, d, e, f, g] => self
+                    .0
+                    .i((
+                        a.to_owned(),
+                        b.to_owned(),
+                        c.to_owned(),
+                        d.to_owned(),
+                        e.to_owned(),
+                        f.to_owned(),
+                        g.to_owned(),
+                    ))
+                    .map_err(wrap_err)?,
+                _ => return Err(PyTypeError::new_err("Only 7 dimensions are supported")),
+            };
```

Comment:
Yeah you are right, using `.narrow(...)` would probably be better here. But couldn't we also use `.index(&[TensorIndexer])` which already implements all the narrows correctly for us if we made `index` and `TensorIndexer` publicly accessible? The py03 wrapper would basically re-implement exactly the same logic as the `index` function and i would prefer to not duplicate the code. 🤔

Reply:
I feel that it will be a lot simpler than craming this in `TensorIndexer`, in particular you will be able to skip totally dimensions on which no filter is done, no error propagation etc, I would hope for it to be very short, like 20 loc or so (but obivously I haven't tried). If you want we can go ahead with your version with all this duplication and I'll tweak it to use `narrow` down the line.

Reply:
What i tried is introducing a simple enum to parse the python slices into: 
```
enum Indexer {
    Index(usize),
    Slice(usize, usize),
    // TODO: Handle Ellipsis
}
```

And then simply narrow the tensor down similar to how you already did it in `.index()`:
```
let mut x = self.0.clone();
        let mut current_dim = 0;

        for (i, indexer) in indexers.iter().enumerate() {
            x = match indexer {

                Indexer::Index(n) => x.narrow(current_dim, *n, 1)?.squeeze(current_dim)?,
                Indexer::Slice(start, stop) => {
                    let out = x.narrow(current_dim, *start, stop.saturating_sub(*start))?;
                    current_dim += 1;
                    out
                }
            }
        }
```

This gets rid of most of the duplication but i basically just reimplemented the `.index()` function. Don't know if this approach would be better 🤔

Reply:
If it's 10 line like this, I think it's ok to duplicate, and it can even optimize for some things that the other version cannot do. The other version tries to be rust idiomatic and compatible with the rust restrictions whereas here you could support negative indexes directly in the indexer rather than doing two steps.

Reply:
Yup, Ellipsis ('...') and negative indexing is much simpler to implement this way. I Updated the implementation and added some more unit-tests for Ellipsis support.